src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -1373,7 +1373,7 @@ static QVariant fcnLength( const QVariantList &values, const QgsExpressionContex
     if ( geom.type() != QgsWkbTypes::LineGeometry )
       return QVariant();
 
-    return QVariant( geom.length() );
+    return QVariant( geom.constGet()->length() );
```

Comment:
@nyalldawson what about this?

Reply:
> Is this different behaviour needed by design, or should it be fixed?

I'd err on leaving it unchanged without a compelling reason, just to avoid risk of project breakage